etna/ciim/models.py

```diff
@@ -1,237 +1,77 @@
-        aren't supported and will raise an UnsupportedSlice error:
-
-        - Negative keys to access results from the end of result set [-1]
-        - Fetching all results [:]
-        - Fetches with a step [0:1:2]
-
-        https://docs.python.org/3/reference/datamodel.html#object.__getitem__
-        """
-
-        if key == slice(None, None, None):
-            raise UnsupportedSlice(
-                "Slicing to return all records ([:]) is not supported"
-            )
-
-        # Key can be an int (in the case of positional fetches [0]) or a slice.
-        # Convert int to slice to make comparisons easier.
-        if isinstance(key, int):
-            key = slice(key, None, None)
-
-        if key.step:
-            raise UnsupportedSlice("Slicing with step is not supported")
-
-        if key.start < 0:
-            raise UnsupportedSlice("Slicing with negative index not supported")
-
-        start = key.start
-        stop = key.stop or 0
-        count = max(1, stop - start)
-        is_single_item_fetch = key.start is not None and not key.stop
-
-        if is_single_item_fetch:
-            try:
-                result = self._results[key.start]
-            except (ResultPending, IndexError):
-                self._fetch(offset=key.start, size=self.BATCH_SIZE)
-                result = self._results[key.start]
-
-            transformed_result = self.model.transform(result)
-            return self.model(**transformed_result)
-
-        self._fetch(offset=key.start, size=count)
-        translated_results = [self.model.transform(r) for r in self._results[key]]
-        return [self.model(**r) for r in translated_results]
-
-    def __len__(self):
-        """Support len()"""
-        return self.count()
-
-    def count(self):
-        self._fetch(offset=0, size=0)
-        return self._response["hits"]["total"]["value"]
-
-    def first(self):
-        """Return the first item in resultset without or None if nothing is found."""
-        try:
-            return self[0]
-        except IndexError:
-            return None
-
-
-class FetchQuery(Query):
-    """Support queries to Kong's /fetch method.
-
-    Once configured via SearchManager.get, this class is responsible for
-    fetching and transforming responses into model instances.
-    """
-
-    def __init__(self, model):
-        self.model = model
-
-    def get(self, iaid=None, reference_number=None, expand=False):
-        response = self.client.fetch(iaid=iaid, id=reference_number, expand=expand)
-
-        results = response["hits"]["hits"]
-        results_count = response["hits"]["total"]["value"]
-        if results_count == 0:
+    def search(self, **kwargs) -> tuple[int, list[APIModel]]:
+        """Make request to /search and transform the response."""
+        response = self.client.search(**kwargs)
+        count, hits = self._extract_hits_and_count(response)
+        return count, [self.model.from_api_response(h) for h in hits]
```

Comment:
Maybe we can do  : 
 if count == self.__RecordNotFound: """this could be a const private member"""
            return count, []
since we know the count result now, so to  avoid executing this line :  [self.model.from_api_response(h) for h in hits]

Reply:
@mozaki1 was this comment meant for elsewhere?

Reply:
Thanks for explaining on slack @mozaki1. So, the proposal as for something like this?

```suggestion
        if not count:
            return 0, []
        return count, [self.model.from_api_response(h) for h in hits]
```

Reply:
Yes , looks good